New module: win_zip

hiyokotaisa opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

Add win_zip module to fix issue #235 

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
- win_zip

jborean93 commented on changelogs/fragments/235-add-win_zip.yml:
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
```
Changelog fragments for new modules aren't needed. It is automatically added when the changelog is generated.

hiyokotaisa replied:
Thanks! I will remove it then

jborean93 commented on plugins/modules/win_zip.ps1:
```diff
@@ -0,0 +1,57 @@
+# Copyright: (c) 2021, Kento Yagisawa <thel.vadam2485@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#Requires -Module Ansible.ModuleUtils.Legacy
+
+$ErrorActionPreference = "Stop"
+
+$params = Parse-Args $args -supports_check_mode $true
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
+$src = Get-AnsibleParam -obj $params -name "src" -type "path" -failifempty $true
+$dest = Get-AnsibleParam -obj $params -name "dest" -type "path" -failifempty $true
+$force = Get-AnsibleParam -obj $params -name "overwrite" -type "bool" -default $false
+
+$result = @{
+    changed = $false
+    dest = $dest -replace '\$',''
+    src = $src -replace '\$',''
+}
```
We should try and use the new module wrapper for this instead of the legacy one. There are numerous benefits you get from using that over the older style

```suggestion
#AnsibleRequires -CSharpUtil Ansible.Basic

$spec = @{
    options = @{
        src = @{ type = 'path'; required = $true }
        dest = @{ type = 'path'; required = $true }
        overwrite = @{ type = 'bool'; default = $false }
    }
    supports_check_mode = $true
}
$module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)

$src = $module.Params.src
$dest = $module.Params.dest
$force = $module.Params.overwrite
```

Some of the features are

* Automatic module invocation details when run with `-vvv`
    * You no longer need to, or should, return `src` and `dest` in your return value
* Better option validation and error messages around that
* Event logging when running the modules

Some things you do need to change with the newer format

* `Fail-Json` becomes `$module.FailJson("My message here")`
* `Exit-Json` become `$module.ExitJson()`
* Getting check mode is now `$module.CheckMode
* Settings results is now done on `$module.Results.foo = 'bar'`, e.g. `$module.Results.changed = $true`

hiyokotaisa replied:
Okay I will rewrite the module with the new module wrapper

jborean93 commented on plugins/modules/win_zip.ps1:
```diff
@@ -0,0 +1,57 @@
+        }
+        $result.changed = $true
+    } catch {
+        Exit-Json -obj $result
```
This will make any errors when you go to compress the zip. You may as well just remove the `try { } catch { }` altogether as you are catching it when calling `Compress-Zip`.

hiyokotaisa replied:
yeah that will work like black hole... I will remove it thanks!

jborean93 commented on plugins/modules/win_zip.ps1:
```diff
@@ -0,0 +1,57 @@
+        Fail-Json -obj $result -message "There is an unexpected error during the zip compression process."
+    }
+
+Exit-Json $result
```
When using the new module wrapper this would now become

```suggestion
$module.ExitJson()
```

jborean93 commented on plugins/modules/win_zip.ps1:
```diff
@@ -0,0 +1,57 @@
+try {
+    Compress-Zip -src $src -dest $dest
+    } catch {
+        Fail-Json -obj $result -message "There is an unexpected error during the zip compression process."
```
It's better if you were to just let the exception go uncaught. We will pick it up automatically and generate a nice error message with a stack trace for you. Unless there's a pressing need to catch specific errors I recommend you don't.

hiyokotaisa replied:
yeah that was unnecessary try catch sentence... thank you for the advice!

jborean93 commented on plugins/modules/win_zip.py:
```diff
@@ -0,0 +1,69 @@
+      - Destination path of zip file (provide absolute path of zip file on the target node).
+    type: path
+    required: yes
+  overwrite:
```
Personally I don't think we should have this. If someone truly wants to always compress the src then they should have a win_file task beforehand to delete it like

```yaml
- win_file:
    path: C:\my\file.zip
    state: absent
```

This allows us to slowly add idempotency to the module that's a lot smarter than just checking if the dest file exists or not.

hiyokotaisa replied:
Yeah, I will remove `overwrite` option then.

jborean93 commented on plugins/modules/win_zip.py:
```diff
@@ -0,0 +1,69 @@
+    type: bool
+    default: false
+notes:
+- The maximum file size is 2GB because there's a limitation of the underlying API.
```
Technically if you use the pure .NET way you can bypass this limit. If you are interested in having that in the module let me know and I can share the details. The other benefit is you aren't limited to just PowerShell v5. The .NET way is possible if you have .NET Framework v4.5 installed.

hiyokotaisa replied:
Thank you for the advice! I will rewrite those with .NET way.

jborean93 commented on plugins/modules/win_zip.py:
```diff
@@ -0,0 +1,69 @@
+'''
+
+RETURN = r'''
+dest:
```
Personally I don't think you should return these, use the new module wrapper and it's done automatically and doesn't need to be documented.

hiyokotaisa replied:
Sure, I will remove it

jborean93 reviewed (changes requested):
Thank you very much for adding the tests, it is greatly appreciated. I've added a few comments on the module itself which I feel like we should implement before merging it in. If you have any questions feel free to let me know.

jborean93 commented on plugins/modules/win_zip.ps1:
```diff
@@ -0,0 +1,59 @@
+        Add-Type -AssemblyName System.IO.Compression.FileSystem
+
+        If (Test-Path -LiteralPath $src -PathType Container) {
+            $zip = [System.IO.Compression.ZipFile]::CreateFromDirectory($src, $dest, $compressionLevel, $true)
```
It would be nice to set the last boolean depending is `$src` ends with `\*` or not. Basically if someone does `src: C:\dir\*` then you don't want to include the base directory (`$false`). If it is just `src: C:\dir` then you want to include the base directory (`$true`).

A minor nit is that `CreateFromDirectory` is a void method and doesn't return anything. You don't need the `$zip =`

jborean93 commented on plugins/modules/win_zip.ps1:
```diff
@@ -0,0 +1,59 @@
+        If (Test-Path -LiteralPath $src -PathType Container) {
+            $zip = [System.IO.Compression.ZipFile]::CreateFromDirectory($src, $dest, $compressionLevel, $true)
+        } Else {
+            $zip = [System.IO.Compression.ZipFile]::Open($dest, 'Update')
+            [System.IO.Compression.ZipFileExtensions]::CreateEntryFromFile($zip, $src, $srcFile, $compressionLevel)
+            $zip.Dispose()
```
You will want to wrap this in a try finally to ensure the zip is disposed even on an error.

```suggestion
            $zip = [System.IO.Compression.ZipFile]::Open($dest, 'Update')
            try {
                [System.IO.Compression.ZipFileExtensions]::CreateEntryFromFile($zip, $src, $srcFile, $compressionLevel)
            } finally {
                $zip.Dispose()
            }
```

jborean93 commented on tests/integration/targets/win_zip/tasks/main.yml:
```diff
@@ -0,0 +1,113 @@
+    state: directory
+
+- name: create test files for CI
+  ansible.builtin.win_shell: 'New-Item -Force -Path C:\testdir\src\test.txt -Value "This is a test file."' 
```
You want to use `Set-Content` here instead of `New-Item`.

hiyokotaisa commented:
Thank you for the review @jborean93 ! :smile: 
I modified the module according to your comment, CI had been failed because of `New-Item` cmdlet but it seems okay after I changed it to `Set-Content`.

jborean93 commented on plugins/modules/win_zip.ps1:
```diff
@@ -49,11 +52,11 @@ try {
             $zip = [System.IO.Compression.ZipFile]::Open($dest, 'Update')
             try {
-                [System.IO.Compression.ZipFileExtensions]::CreateEntryFromFile($zip, $src, $srcFile, $compressionLevel)
+                [void][System.IO.Compression.ZipFileExtensions]::CreateEntryFromFile($zip, $src, $srcFile, $compressionLevel)
```
`CreateEntryFromFile` outputs data, to avoid it clashing with the module return json and showing an error we just void it out.

jborean93 commented on tests/integration/targets/win_zip/tasks/main.yml:
```diff
@@ -1,33 +1,54 @@
-  ansible.windows.win_file:
-    path: C:\testdir\
-    state: absent
+- set_fact:
```
This is to just expand the tests a bit more and validate it zipped the files as we expected.

jborean93 commented on tests/integration/targets/win_zip/meta/main.yml:
```diff
@@ -0,0 +1,2 @@
+dependencies:
+- setup_remote_tmp_dir
```
This handles the creation and deletion of temp dirs in tests, you can access them with `{{ remote_tmp_dir }}` as done in `defaults/main.yml`.

jborean93 commented on tests/integration/targets/win_zip/defaults/main.yml:
```diff
@@ -0,0 +1,2 @@
+win_zip_name: .ÅÑŚÌβŁÈ [$!@^&test(;)] 👋
```
Using some non-ASCII characters here to test out some edge cases when it comes to Windows. This ensures that the module is able to handle these non-ASCII chars and verifies the UTF-8 encoding we specify is actually in affect.

jborean93 commented:
Ignoring Zuul check as it's a known failure and unrelated to any changes here.
